src/parser/ion_1_0/text/mod.rs

```diff
@@ -0,0 +1,2614 @@
+/// TYPE
+///     : 'bool'
+///     | 'int'
+///     | 'float'
+///     | 'decimal'
+///     | 'timestamp'
+///     | 'symbol'
+///     | 'string'
+///     | 'clob'
+///     | 'blob'
+///     | 'list'
+///     | 'sexp'
+///     | 'struct'
+///     ;
+fn take_type(i: &str) -> IonResult<&str, &str> {
+    alt((
+        tag("null"),
+        tag("bool"),
+        tag("int"),
+        tag("float"),
+        tag("decimal"),
+        tag("timestamp"),
+        tag("string"),
+        tag("symbol"),
+        tag("blob"),
+        tag("clob"),
+        tag("struct"),
+        tag("list"),
+        tag("sexp"),
+    ))(i)
+}
+
+///
+/// Encoding Section: Ion Bool
+///
+
+/// Encoding: BOOL
+/// BOOL
+///     : 'true'
+///     | 'false'
+///     ;
+fn take_bool(i: &str) -> IonResult<&str, bool> {
+    alt((value(true, tag("true")), value(false, tag("false"))))(i)
+}
+
+///
+/// Encoding Section: Ion TextTimestamp
+///
+
+/// Encoding: fragment TIMESTAMP
+/// TIMESTAMP
+///     : DATE ('T' TIME?)?
+///     | YEAR '-' MONTH 'T'
+///     | YEAR 'T'
+///     ;
+fn take_timestamp(i: &str) -> IonResult<&str, ion::Timestamp> {
+    let (i, timestamp) = map(
+        alt((
+            map_res(
+                pair(take_date, opt(preceded(one_of("tT"), opt(take_time)))),
+                |((year, month, day), maybe_time)| match ion::Date::day(year, month, day) {
+                    Ok(date) => {
+                        if let Some(time) = maybe_time {
+                            Ok(ion::TextTimestamp::new(date, time))
+                        } else {
+                            Ok(ion::TextTimestamp::new(date, None))
+                        }
+                    }
+                    Err(e) => Err(e),
+                },
+            ),
+            // Timestamps which consist of only a year or year and month must be terminated by t or T.
+            map(
+                terminated(
+                    separated_pair(take_year, char('-'), take_month),
+                    one_of("tT"),
+                ),
+                |(year, month)| ion::TextTimestamp::new(ion::Date::Month { year, month }, None),
+            ),
+            map(terminated(take_year, one_of("tT")), |year| {
+                ion::TextTimestamp::new(ion::Date::Year { year }, None)
+            }),
+        )),
+        ion::Timestamp::Text,
+    )(i)?;
+
+    Ok((i, timestamp))
+}
+
+/// fragment
+/// DATE
+///     : YEAR '-' MONTH '-' DAY
+///     ;
+fn take_date(i: &str) -> IonResult<&str, (i32, u8, u8)> {
+    let (i, (y, (m, d))) = pair(
+        take_year,
+        pair(
+            preceded(char('-'), take_month),
+            preceded(char('-'), take_day),
+        ),
+    )(i)?;
+
+    Ok((i, (y, m, d)))
+}
+
+/// fragment
+/// YEAR
+///     : '000'                     [1-9]
+///     | '00'            [1-9]     DEC_DIGIT
+///     | '0'   [1-9]     DEC_DIGIT DEC_DIGIT
+///     | [1-9] DEC_DIGIT DEC_DIGIT DEC_DIGIT
+///     ;
+fn take_year(i: &str) -> IonResult<&str, i32> {
```

Comment:
I think ISO supports arbitrary years, so I would expect `take_year` to parse as a `BigInt`. I assume it's an `i32` so that it can be checked with a `time::Date`?

Reply:
Yeah, that's why. My read of the W3C note implies that only a four digit year is supported (YYYY).

Reply:
I took a look at the W3C note and I think you're right.